File: compiler/frontend/src/org/jetbrains/jet/lang/resolve/MutablePackageFragmentProvider.java
package org.jetbrains.jet.lang.resolve;

import com.google.common.collect.Maps;
import com.google.common.collect.Multimap;
import com.intellij.util.containers.ContainerUtil;
import com.intellij.util.containers.MultiMap;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.jet.lang.descriptors.ModuleDescriptor;
import org.jetbrains.jet.lang.descriptors.PackageFragmentDescriptor;
import org.jetbrains.jet.lang.descriptors.PackageFragmentProvider;
import org.jetbrains.jet.lang.descriptors.impl.MutablePackageFragmentDescriptor;
import org.jetbrains.jet.lang.resolve.name.FqName;

import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public class MutablePackageFragmentProvider implements PackageFragmentProvider {
    private final ModuleDescriptor module;

    private final Map<FqName, MutablePackageFragmentDescriptor> fqNameToPackage = Maps.newHashMap();
    private final MultiMap<FqName, FqName> subPackages = MultiMap.create();

    public MutablePackageFragmentProvider(@NotNull ModuleDescriptor module) {
        this.module = module;
        fqNameToPackage.put(FqName.ROOT, new MutablePackageFragmentDescriptor(module, FqName.ROOT));
    }

    @NotNull
    @Override
    public List<PackageFragmentDescriptor> getPackageFragments(@NotNull FqName fqName) {
        return ContainerUtil.<PackageFragmentDescriptor>createMaybeSingletonList(fqNameToPackage.get(fqName));
    }

    @NotNull
    @Override
    public Collection<FqName> getSubPackagesOf(@NotNull FqName fqName) {
        return subPackages.get(fqName);
    }

    @NotNull
    public MutablePackageFragmentDescriptor getOrCreateFragment(@NotNull FqName fqName) {
        if (!fqNameToPackage.containsKey(fqName)) {
            FqName parent = fqName.parent();
            getOrCreateFragment(parent); // assure that parent exists

            fqNameToPackage.put(fqName, new MutablePackageFragmentDescriptor(module, fqName));
            subPackages.putValue(parent, fqName);
        }

        return fqNameToPackage.get(fqName);
    }

    @NotNull
    public ModuleDescriptor getModule() {
        return module;
    }

    @NotNull
    public Collection<MutablePackageFragmentDescriptor> getAllFragments() {
        return fqNameToPackage.values();
    }
}

File: compiler/frontend/src/org/jetbrains/kotlin/resolve/MutablePackageFragmentProvider.kt
/*
 * Copyright 2010-2015 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.kotlin.resolve

import com.intellij.util.containers.ContainerUtil
import com.intellij.util.containers.MultiMap
import org.jetbrains.kotlin.descriptors.ModuleDescriptor
import org.jetbrains.kotlin.descriptors.PackageFragmentDescriptor
import org.jetbrains.kotlin.descriptors.PackageFragmentProvider
import org.jetbrains.kotlin.descriptors.impl.MutablePackageFragmentDescriptor
import org.jetbrains.kotlin.name.FqName
import java.util.HashMap
import org.jetbrains.kotlin.name.Name

public class MutablePackageFragmentProvider(public val module: ModuleDescriptor) : PackageFragmentProvider {

    private val fqNameToPackage = HashMap<FqName, MutablePackageFragmentDescriptor>()
    private val subPackages = MultiMap.create<FqName, FqName>()

    ;{
        fqNameToPackage.put(FqName.ROOT, MutablePackageFragmentDescriptor(module, FqName.ROOT))
    }

    override fun getPackageFragments(fqName: FqName)
            = ContainerUtil.createMaybeSingletonList<PackageFragmentDescriptor>(fqNameToPackage.get(fqName))

    override fun getSubPackagesOf(fqName: FqName, nameFilter: (Name) -> Boolean) = subPackages[fqName]

    public fun getOrCreateFragment(fqName: FqName): MutablePackageFragmentDescriptor {
        if (!fqNameToPackage.containsKey(fqName)) {
            val parent = fqName.parent()
            getOrCreateFragment(parent) // assure that parent exists

            fqNameToPackage.put(fqName, MutablePackageFragmentDescriptor(module, fqName))
            subPackages.putValue(parent, fqName)
        }

        return fqNameToPackage[fqName]
    }

    public fun getAllFragments(): Collection<MutablePackageFragmentDescriptor> = fqNameToPackage.values()
}
